stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java

```diff
@@ -679,38 +682,24 @@ public static BStruct createNewEntity(Context context, BStruct struct) {
         return entity;
     }
 
-    /**
-     * Set connection content-encoding headers to transport message.
-     *
-     * @param context         ballerina context.
-     * @param outboundMessage transport message.
-     */
-    public static void setKeepAliveAndCompressionHeaders(Context context, HTTPCarbonMessage outboundMessage) {
-        HttpService service = (HttpService) context.getProperty(HttpConstants.HTTP_SERVICE);
-
-        if (!service.isCompressionEnabled()) {
-            outboundMessage.setHeader(HttpHeaderNames.CONTENT_ENCODING.toString(),
-                    Constants.HTTP_TRANSFER_ENCODING_IDENTITY);
-        }
-
-        if (service.isKeepAlive()) {
-            outboundMessage.setHeader(HttpConstants.CONNECTION_HEADER,
-                    HttpConstants.HEADER_VAL_CONNECTION_KEEP_ALIVE);
-            return;
-        }
-        outboundMessage.setHeader(HttpConstants.CONNECTION_HEADER, HttpConstants.HEADER_VAL_CONNECTION_CLOSE);
-    }
-
-    public static void setCompressionHeaders(Context context, HTTPCarbonMessage outboundMessage) {
+    static void setCompressionHeaders(Context context, HTTPCarbonMessage requestMsg, HTTPCarbonMessage
+            outboundResponseMsg) {
         Service serviceInstance = BLangConnectorSPIUtil.getService(context.getProgramFile(),
-                                                                   context.getServiceInfo().getType());
+                context.getServiceInfo().getType());
         Annotation configAnnot = getServiceConfigAnnotation(serviceInstance, PROTOCOL_PACKAGE_HTTP);
-
         if (configAnnot != null) {
-            boolean isCompressionEnabled = configAnnot.getValue().getBooleanField(ANN_CONFIG_ATTR_COMPRESSION_ENABLED);
-            if (isCompressionEnabled) {
-                outboundMessage.setHeader(HttpHeaderNames.CONTENT_ENCODING.toString(),
-                                          Constants.HTTP_TRANSFER_ENCODING_IDENTITY);
+            String compressionValue = configAnnot.getValue().getEnumField(ANN_CONFIG_ATTR_COMPRESSION);
+            String acceptEncodingValue = requestMsg.getHeaders().get(HttpHeaderNames.ACCEPT_ENCODING);
+            String contentEncoding = outboundResponseMsg.getHeaders().get(HttpHeaderNames.CONTENT_ENCODING);
+            if (contentEncoding == null) {
+                if (ALWAYS.equalsIgnoreCase(compressionValue)) {
+                    if (acceptEncodingValue == null || HTTP_TRANSFER_ENCODING_IDENTITY.equals(acceptEncodingValue)) {
```

Comment:
Shall we reduce the nested if clauses here?
We can check the condition and exit at the top, without going to nested if clauses.